src/error_handling.c

```diff
@@ -0,0 +1,259 @@
+
+void
+__utilities_reset_error(utilities_error_state_t ** error_state_ptr_ptr);
+
+void
+utilities_set_error_state(
+  const char * error_string,
+  const char * file,
+  size_t line_number,
+  utilities_allocator_t allocator)
+{
+#ifdef UTILITIES_THREAD_LOCAL_PTHREAD
+  utilities_error_state_t * __utilities_error_state =
+    (utilities_error_state_t *)pthread_getspecific(__utilities_error_state_key);
+  char * __utilities_error_string = (char *)pthread_getspecific(__utilities_error_string_key);
+#endif
+  utilities_error_state_t * old_error_state = __utilities_error_state;
+#if UTILITIES_REPORT_ERROR_HANDLING_ERRORS
+  const char * old_error_string = utilities_get_error_string_safe();
+#endif
+  __utilities_error_state = (utilities_error_state_t *)allocator.allocate(
+    sizeof(utilities_error_state_t), allocator.state);
+  if (!__utilities_error_state) {
+#if UTILITIES_REPORT_ERROR_HANDLING_ERRORS
+    // utilities_allocate failed, but fwrite might work?
+    SAFE_FWRITE_TO_STDERR(
+      "[rmw|error_handling.c:" UTILITIES_STRINGIFY(__LINE__)
+      "] failed to allocate memory for the error state struct\n");
+#endif
+    return;
+  }
+  __utilities_error_state->allocator = allocator;
+
+#ifdef UTILITIES_THREAD_LOCAL_PTHREAD
+  pthread_setspecific(__utilities_error_state_key, __utilities_error_state);
+#endif
+  size_t error_string_length = strlen(error_string);
+  // the memory must be one byte bigger to store the NULL character
+  __utilities_error_state->message =
+    (char *)allocator.allocate(error_string_length + 1, allocator.state);
+  if (!__utilities_error_state->message) {
+#if UTILITIES_REPORT_ERROR_HANDLING_ERRORS
+    // malloc failed, but fwrite might work?
+    SAFE_FWRITE_TO_STDERR(
+      "[rmw|error_handling.c:" UTILITIES_STRINGIFY(__LINE__)
+      "] failed to allocate memory for the error message in the error state struct\n");
+#endif
+    utilities_reset_error();  // This will free any allocations done so far.
+    return;
+  }
+  // Cast the const away to set ->message initially.
+#ifndef _WIN32
+  snprintf((char *)__utilities_error_state->message, error_string_length + 1, "%s", error_string);
+#else
+  auto retcode = strcpy_s(
+    (char *)__utilities_error_state->message, error_string_length + 1, error_string);
+  if (retcode) {
+#if UTILITIES_REPORT_ERROR_HANDLING_ERRORS
+    SAFE_FWRITE_TO_STDERR(
+      "[rmw|error_handling.c:" UTILITIES_STRINGIFY(__LINE__)
+      "] failed to copy error message in the error state struct\n");
+#endif
+  }
+#endif
+  __utilities_error_state->file = file;
+  __utilities_error_state->line_number = line_number;
+  if (__utilities_error_is_set(old_error_state)) {
+#if UTILITIES_REPORT_ERROR_HANDLING_ERRORS
+    // Only warn of overwritting if the new error string is different from the old ones.
+    if (error_string != old_error_string && error_string != old_error_state->message) {
+      fprintf(
```

Comment:
This error message uses formatting (like `%s`), fwrite cannot do that. I only fallback to fwrite when memory allocation fails, because apparently it can still work on some systems even when malloc fails.